File: Frontend/src/app/services/training.service.ts
import { Injectable } from '@angular/core';
import { Training } from '../models/training.model';
import axios from "axios";

@Injectable({
  providedIn: 'root'
})
export class TrainingService {
  private baseUrl = 'http://localhost:8080/api/trainings';

  async getTrainings(): Promise<Training[]> {
    try {
      const response = await axios.get<Training[]>(this.baseUrl, { withCredentials: true });
      return response.data;
    } catch (error) {
      console.error('Error fetching trainings', error);
      throw error;
    }
  }

  async getTraining(id: number): Promise<Training> {
    try {
      const response = await axios.get<Training>(`${this.baseUrl}/${id}`, { withCredentials: true });
      return response.data;
    } catch (error) {
      console.error('Error fetching training', error);
      throw error;
    }
  }

  async createTraining(training: Training): Promise<Training> {
    try {
      const response = await axios.post<Training>(this.baseUrl, training, { withCredentials: true });
      return response.data;
    } catch (error) {
      console.error('Error creating training', error);
      throw error;
    }
  }

  async updateTraining(id: number, training: Training): Promise<Training> {
    try {
      const response = await axios.put<Training>(`${this.baseUrl}/${id}`, training, { withCredentials: true });
      return response.data;
    } catch (error) {
      console.error('Error updating training', error);
      throw error;
    }
  }

  async deleteTraining(id: number): Promise<void> {
    try {
      await axios.delete(`${this.baseUrl}/${id}`, { withCredentials: true });
    } catch (error) {
      console.error('Error deleting training', error);
      throw error;
    }
  }

  async toggleEnrollment(trainingId: number, userId: number): Promise<void> {
    try {
      await axios.post(`${this.baseUrl}/${trainingId}/toggle-enrollment`, null, {
        params: { userId },
        withCredentials: true,
      });
      console.log('Enrollment toggled');
    } catch (error) {
      console.error('Error toggling enrollment', error);
      throw error;
    }
  }

  async getUserTrainings(userId: number): Promise<Training[]> {
    try {
      const response = await axios.get<Training[]>(`${this.baseUrl}/user/${userId}`, { withCredentials: true });
      return response.data;
    } catch (error) {
      console.error('Error fetching user trainings', error);
      throw error;
    }
  }
}

File: dev/src/main/java/com/crm/dev/service/TrainingService.java
package com.crm.dev.service;

import com.crm.dev.models.Training;
import com.crm.dev.models.User;
import com.crm.dev.models.UserTraining;
import com.crm.dev.repository.TrainingRepository;
import com.crm.dev.repository.UserRepository;
import com.crm.dev.repository.UserTrainingRepository;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class TrainingService {

    private static final Logger logger = LoggerFactory.getLogger(TrainingService.class);

    @Autowired
    private TrainingRepository trainingRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private UserTrainingRepository userTrainingRepository;

    public List<Training> getAllTrainings() {
        return trainingRepository.findAll();
    }

    public Training getTrainingById(Long id) {
        return trainingRepository.findById(id).orElseThrow(() -> new RuntimeException("Training not found"));
    }

    public Training createTraining(Training training) {
        logger.info("Creating new training with title: {}", training.getTitle());
        return trainingRepository.save(training);
    }

    public Training updateTraining(Long id, Training trainingDetails) {
        Training training = trainingRepository.findById(id).orElseThrow(() -> new RuntimeException("Training not found"));
        logger.info("Updating training with ID: {}", id);
        training.setTitle(trainingDetails.getTitle());
        training.setDescription(trainingDetails.getDescription());
        training.setDate(trainingDetails.getDate());
        return trainingRepository.save(training);
    }

    public void deleteTraining(Long id) {
        Training training = trainingRepository.findById(id).orElseThrow(() -> new RuntimeException("Training not found"));
        logger.info("Deleting training with ID: {}", id);
        trainingRepository.delete(training);
    }

    @Transactional
    public void toggleEnrollment(Long trainingId, Long userId) {
        logger.info("Toggling enrollment for user with ID: {} in training with ID: {}", userId, trainingId);

        // Vérifie si l'utilisateur est déjà inscrit à la formation
        if (userTrainingRepository.existsByUserIdAndTrainingId(userId, trainingId)) {
            logger.info("User with ID: {} is already enrolled in training with ID: {}. Removing enrollment.", userId, trainingId);
            userTrainingRepository.deleteByUserIdAndTrainingId(userId, trainingId);
            logger.info("Enrollment removed for user with ID: {} in training with ID: {}", userId, trainingId);
        } else {
            // Cherche l'utilisateur et la formation
            User user = userRepository.findById(userId).orElseThrow(() -> {
                logger.error("User not found with ID: {}", userId);
                return new RuntimeException("User not found");
            });
            Training training = trainingRepository.findById(trainingId).orElseThrow(() -> {
                logger.error("Training not found with ID: {}", trainingId);
                return new RuntimeException("Training not found");
            });

            logger.info("User with ID: {} is not enrolled in training with ID: {}. Adding enrollment.", userId, trainingId);

            // Crée une nouvelle inscription
            UserTraining userTraining = new UserTraining();
            userTraining.setUser(user);
            userTraining.setTraining(training);
            userTrainingRepository.save(userTraining);
            logger.info("Enrollment added for user with ID: {} in training with ID: {}", userId, trainingId);
        }
    }

    public List<Training> getUserTrainings(Long userId) {
        List<UserTraining> userTrainings = userTrainingRepository.findByUserId(userId);
        return userTrainings.stream()
                .map(UserTraining::getTraining)
                .collect(Collectors.toList());
    }


}
